src/networking.c

```diff
@@ -1751,19 +1773,19 @@ void clearClientConnectionState(client *c) {
     c->flag.no_evict = 0;
 }
 
-void freeClient(client *c) {
+/* Free the client structure and all the data associated with it.
+ * Returns 0 if the client was not freed immediately, but scheduled for
+ * asynchronous freeing, and 1 if the client was freed immediately. */
+int freeClient(client *c) {
     listNode *ln;
 
     /* If a client is protected, yet we need to free it right now, make sure
      * to at least use asynchronous freeing. */
-    if (c->flag.protected || c->flag.protected_rdb_channel) {
+    if (c->flag.protected || c->flag.protected_rdb_channel || clientIOInProgress(c)) {
```

Comment:
```suggestion
    if (c->flag.protected || c->flag.protected_rdb_channel || clientHandlingThreadedIO(c)) {
```
Something should indicate this is on the thread.

Reply:
Done